File: bindings/c++/examples/sign_example.cpp
/**
 * @file sign_example.cpp
 * @brief QUAC 100 C++ SDK - Digital Signature Example
 * @copyright Copyright © 2025 Dyber, Inc. All Rights Reserved.
 *
 * Demonstrates ML-DSA (Dilithium) post-quantum digital signatures.
 */

#include <quac100/quac100.hpp>
#include <iostream>
#include <iomanip>
#include <chrono>

using namespace quac100;

void printBytes(const std::string &label, const Bytes &data, size_t maxLen = 32)
{
    std::cout << label << " (" << data.size() << " bytes): ";
    std::cout << utils::toHex(Bytes(data.begin(),
                                    data.begin() + std::min(data.size(), maxLen)));
    if (data.size() > maxLen)
        std::cout << "...";
    std::cout << "\n";
    std::cout.flush();
}

std::string formatDuration(double ms)
{
    std::ostringstream oss;
    if (ms < 1.0)
    {
        oss << std::fixed << std::setprecision(2) << (ms * 1000.0) << " us";
    }
    else if (ms < 1000.0)
    {
        oss << std::fixed << std::setprecision(2) << ms << " ms";
    }
    else
    {
        oss << std::fixed << std::setprecision(2) << (ms / 1000.0) << " s";
    }
    return oss.str();
}

int main()
{
    std::cout << "QUAC 100 C++ SDK - Digital Signature Example\n";
    std::cout << "=============================================\n\n";
    std::cout.flush();

    try
    {
        std::cout << "Initializing library...\n";
        std::cout.flush();
        Library lib;

        std::cout << "Opening device...\n";
        std::cout.flush();
        auto device = lib.openFirstDevice();
        std::cout << "Device opened successfully.\n\n";
        std::cout.flush();

        // Get algorithm parameters (use instance method via device.sign())
        std::cout << "ML-DSA Algorithm Parameters:\n";
        std::cout << "----------------------------\n";
        std::cout.flush();

        for (auto alg : {SignAlgorithm::ML_DSA_44,
                         SignAlgorithm::ML_DSA_65,
                         SignAlgorithm::ML_DSA_87})
        {
            auto params = device.sign().getParams(alg);
            std::cout << params.name << ":\n";
            std::cout << "  Public Key:  " << params.publicKeySize << " bytes\n";
            std::cout << "  Secret Key:  " << params.secretKeySize << " bytes\n";
            std::cout << "  Signature:   " << params.signatureSize << " bytes\n";
            std::cout << "  Security:    Level " << params.securityLevel << "\n\n";
            std::cout.flush();
        }

        // Demonstrate ML-DSA-65 signing
        std::cout << "ML-DSA-65 Digital Signature Demo\n";
        std::cout << "---------------------------------\n\n";
        std::cout.flush();

        // Generate signing key pair
        std::cout << "Generating signing key pair...\n";
        std::cout.flush();
        auto startTime = std::chrono::high_resolution_clock::now();
        auto keys = device.sign().generateKeyPair(SignAlgorithm::ML_DSA_65);
        auto endTime = std::chrono::high_resolution_clock::now();
        double keygenMs = std::chrono::duration<double, std::milli>(endTime - startTime).count();
        std::cout << "Key generation took " << formatDuration(keygenMs) << "\n";
        printBytes("Public Key", keys.publicKey);
        std::cout << "\n";
        std::cout.flush();

        // Sign a message
        std::string messageStr = "This is an important document that needs to be signed.";
        Bytes message(messageStr.begin(), messageStr.end());
        std::cout << "Message: \"" << messageStr << "\"\n\n";
        std::cout.flush();

        std::cout << "Signing message...\n";
        std::cout.flush();
        startTime = std::chrono::high_resolution_clock::now();
        auto signature = device.sign().sign(SignAlgorithm::ML_DSA_65, keys.secretKey, message);
        endTime = std::chrono::high_resolution_clock::now();
        double signMs = std::chrono::duration<double, std::milli>(endTime - startTime).count();
        std::cout << "Signing took " << formatDuration(signMs) << "\n";
        printBytes("Signature", signature);
        std::cout << "\n";
        std::cout.flush();

        // Verify signature
        std::cout << "Verifying signature...\n";
        std::cout.flush();
        startTime = std::chrono::high_resolution_clock::now();
        bool valid = device.sign().verify(SignAlgorithm::ML_DSA_65,
                                          keys.publicKey, message, signature);
        endTime = std::chrono::high_resolution_clock::now();
        double verifyMs = std::chrono::duration<double, std::milli>(endTime - startTime).count();
        std::cout << "Verification took " << formatDuration(verifyMs) << "\n";
        std::cout.flush();

        if (valid)
        {
            std::cout << "[OK] Signature is VALID\n\n";
        }
        else
        {
            std::cout << "[FAIL] Signature is INVALID\n";
            return 1;
        }
        std::cout.flush();

        // Test tamper detection
        std::cout << "Testing Tamper Detection:\n";
        std::cout << "-------------------------\n";
        std::cout.flush();

        // Modify message
        std::string tamperedStr = "This is an important document that needs to be signed!";
        Bytes tamperedMessage(tamperedStr.begin(), tamperedStr.end());
        std::cout << "Tampered: \"" << tamperedStr << "\"\n";
        std::cout.flush();

        bool tamperedValid = false;
        try
        {
            tamperedValid = device.sign().verify(SignAlgorithm::ML_DSA_65,
                                                 keys.publicKey,
                                                 tamperedMessage,
                                                 signature);
        }
        catch (const VerificationException &)
        {
            tamperedValid = false;
        }

        if (!tamperedValid)
        {
            std::cout << "[OK] Tamper detected: Signature rejected for modified message\n\n";
        }
        else
        {
            std::cout << "[FAIL] SECURITY FAILURE: Tampered message accepted!\n";
            return 1;
        }
        std::cout.flush();

        // Batch signing example
        std::cout << "Batch Signing Example:\n";
        std::cout << "----------------------\n";
        std::cout.flush();

        std::vector<Bytes> messages = {
            {'M', 'e', 's', 's', 'a', 'g', 'e', ' ', '1'},
            {'M', 'e', 's', 's', 'a', 'g', 'e', ' ', '2'},
            {'M', 'e', 's', 's', 'a', 'g', 'e', ' ', '3'},
            {'M', 'e', 's', 's', 'a', 'g', 'e', ' ', '4'},
            {'M', 'e', 's', 's', 'a', 'g', 'e', ' ', '5'}};

        std::vector<Bytes> signatures;
        startTime = std::chrono::high_resolution_clock::now();
        for (const auto &msg : messages)
        {
            signatures.push_back(device.sign().sign(SignAlgorithm::ML_DSA_65,
                                                    keys.secretKey, msg));
        }
        endTime = std::chrono::high_resolution_clock::now();
        double batchSignMs = std::chrono::duration<double, std::milli>(endTime - startTime).count();
        std::cout << "Signed " << messages.size() << " messages in "
                  << formatDuration(batchSignMs) << "\n";
        std::cout.flush();

        int validCount = 0;
        startTime = std::chrono::high_resolution_clock::now();
        for (size_t i = 0; i < messages.size(); ++i)
        {
            try
            {
                if (device.sign().verify(SignAlgorithm::ML_DSA_65,
                                         keys.publicKey,
                                         messages[i],
                                         signatures[i]))
                {
                    validCount++;
                }
            }
            catch (...)
            {
            }
        }
        endTime = std::chrono::high_resolution_clock::now();
        double batchVerifyMs = std::chrono::duration<double, std::milli>(endTime - startTime).count();
        std::cout << "Verified " << messages.size() << " signatures in "
                  << formatDuration(batchVerifyMs) << "\n";
        std::cout << "Results: " << validCount << "/" << messages.size() << " valid\n\n";
        std::cout.flush();

        // Performance benchmark
        std::cout << "Performance Benchmark (100 operations):\n";
        std::cout << "---------------------------------------\n";
        std::cout.flush();

        const int iterations = 100;
        Bytes testMsg(256, 0x42);

        // Sign benchmark
        startTime = std::chrono::high_resolution_clock::now();
        for (int i = 0; i < iterations; ++i)
        {
            device.sign().sign(SignAlgorithm::ML_DSA_65, keys.secretKey, testMsg);
        }
        endTime = std::chrono::high_resolution_clock::now();
        double totalMs = std::chrono::duration<double, std::milli>(endTime - startTime).count();
        double signRate = iterations / (totalMs / 1000.0);
        std::cout << "Signing:      " << std::fixed << std::setprecision(0)
                  << signRate << " ops/sec\n";
        std::cout.flush();

        // Verify benchmark
        auto testSig = device.sign().sign(SignAlgorithm::ML_DSA_65, keys.secretKey, testMsg);
        startTime = std::chrono::high_resolution_clock::now();
        for (int i = 0; i < iterations; ++i)
        {
            device.sign().verify(SignAlgorithm::ML_DSA_65, keys.publicKey, testMsg, testSig);
        }
        endTime = std::chrono::high_resolution_clock::now();
        totalMs = std::chrono::duration<double, std::milli>(endTime - startTime).count();
        double verifyRate = iterations / (totalMs / 1000.0);
        std::cout << "Verification: " << std::fixed << std::setprecision(0)
                  << verifyRate << " ops/sec\n\n";
        std::cout.flush();

        std::cout << "Signature example completed successfully!\n";
        std::cout.flush();
    }
    catch (const Exception &e)
    {
        std::cerr << "QUAC Error: " << e.what() << " (code: " << e.codeInt() << ")\n";
        std::cerr.flush();
        return 1;
    }
    catch (const std::exception &e)
    {
        std::cerr << "Standard exception: " << e.what() << "\n";
        std::cerr.flush();
        return 1;
    }

    return 0;
}

File: bindings/java/examples/SignExample.java
/*
 * Copyright © 2025 Dyber, Inc. All Rights Reserved.
 * QUAC 100 Post-Quantum Cryptographic Accelerator - Java SDK
 * 
 * Sign Example - Digital Signature operations
 */
package com.dyber.quac100.examples;

import com.dyber.quac100.*;
import java.nio.charset.StandardCharsets;

/**
 * Example demonstrating ML-DSA (Dilithium) digital signatures.
 */
public class SignExample {

    public static void main(String[] args) {
        System.out.println("=== QUAC 100 Java SDK Signature Example ===\n");

        try (Library lib = Library.getInstance()) {
            Device device = lib.openFirstDevice();
            Sign sign = device.sign();

            // Print algorithm parameters
            System.out.println("ML-DSA Algorithm Parameters:\n");
            for (SignAlgorithm alg : SignAlgorithm.values()) {
                SignParams params = sign.getParams(alg);
                System.out.println("  " + params.getName() + ":");
                System.out.println("    Public Key:  " + params.getPublicKeySize() + " bytes");
                System.out.println("    Secret Key:  " + params.getSecretKeySize() + " bytes");
                System.out.println("    Signature:   " + params.getSignatureSize() + " bytes (max)");
                System.out.println("    Security:    Level " + params.getSecurityLevel());
                System.out.println();
            }

            // Demonstrate signing with ML-DSA-65 (recommended)
            System.out.println("=== ML-DSA-65 Signature Demo ===\n");

            // Generate signing key pair
            System.out.println("Generating signing key pair...");
            long startTime = System.nanoTime();
            KeyPair keys = sign.generateKeyPair65();
            long keyGenTime = System.nanoTime() - startTime;

            System.out.println("  Public key:  " + keys.getPublicKeySize() + " bytes");
            System.out.println("  Secret key:  " + keys.getSecretKeySize() + " bytes");
            System.out.println("  Time:        " + (keyGenTime / 1000.0) + " µs");
            System.out.println();

            // Sign a message
            String message = "This is an important document that needs to be signed.";
            byte[] messageBytes = message.getBytes(StandardCharsets.UTF_8);

            System.out.println("Signing message: \"" + message + "\"");
            startTime = System.nanoTime();
            byte[] signature = sign.sign65(keys.getSecretKey(), messageBytes);
            long signTime = System.nanoTime() - startTime;

            System.out.println("  Signature size: " + signature.length + " bytes");
            System.out.println("  Signature (first 64 bytes): " +
                    Utils.toHex(java.util.Arrays.copyOf(signature, 64)) + "...");
            System.out.println("  Time:           " + (signTime / 1000.0) + " µs");
            System.out.println();

            // Verify the signature
            System.out.println("Verifying signature...");
            startTime = System.nanoTime();
            boolean valid = sign.verify65(keys.getPublicKey(), messageBytes, signature);
            long verifyTime = System.nanoTime() - startTime;

            System.out.println("  Valid:  " + (valid ? "YES ✓" : "NO ✗"));
            System.out.println("  Time:   " + (verifyTime / 1000.0) + " µs");
            System.out.println();

            // Verify tamper detection
            System.out.println("Testing tamper detection...");
            byte[] tamperedMessage = "This is a TAMPERED document that needs to be signed."
                    .getBytes(StandardCharsets.UTF_8);

            boolean tamperedValid = sign.verify65(keys.getPublicKey(), tamperedMessage, signature);
            System.out.println("  Tampered message verification: " +
                    (tamperedValid ? "VALID (BAD!)" : "REJECTED ✓"));

            if (tamperedValid) {
                System.err.println("ERROR: Tampered message should have been rejected!");
                System.exit(1);
            }
            System.out.println();

            // Batch signing demo
            System.out.println("=== Batch Signing Demo ===\n");
            String[] documents = {
                    "Contract Agreement v1.0",
                    "Purchase Order #12345",
                    "Invoice 2025-001",
                    "Shipping Manifest",
                    "Quality Certificate"
            };

            System.out.println("Signing " + documents.length + " documents:\n");
            for (String doc : documents) {
                byte[] docBytes = doc.getBytes(StandardCharsets.UTF_8);
                byte[] docSig = sign.sign65(keys.getSecretKey(), docBytes);
                boolean docValid = sign.verify65(keys.getPublicKey(), docBytes, docSig);
                System.out.println("  \"" + doc + "\"");
                System.out.println("    Sig: " + Utils.toHex(java.util.Arrays.copyOf(docSig, 16)) + "...");
                System.out.println("    Valid: " + (docValid ? "✓" : "✗"));
            }
            System.out.println();

            // Performance benchmark
            System.out.println("=== Performance Benchmark (100 iterations) ===\n");

            int iterations = 100;
            byte[] benchMsg = "Benchmark message for performance testing".getBytes(StandardCharsets.UTF_8);
            long totalKeyGen = 0, totalSign = 0, totalVerify = 0;

            for (int i = 0; i < iterations; i++) {
                startTime = System.nanoTime();
                KeyPair kp = sign.generateKeyPair65();
                totalKeyGen += System.nanoTime() - startTime;

                startTime = System.nanoTime();
                byte[] sig = sign.sign65(kp.getSecretKey(), benchMsg);
                totalSign += System.nanoTime() - startTime;

                startTime = System.nanoTime();
                sign.verify65(kp.getPublicKey(), benchMsg, sig);
                totalVerify += System.nanoTime() - startTime;

                kp.destroy();
            }

            System.out.printf("KeyGen:  %.2f µs/op (%.0f ops/sec)%n",
                    totalKeyGen / 1000.0 / iterations,
                    iterations * 1_000_000_000.0 / totalKeyGen);
            System.out.printf("Sign:    %.2f µs/op (%.0f ops/sec)%n",
                    totalSign / 1000.0 / iterations,
                    iterations * 1_000_000_000.0 / totalSign);
            System.out.printf("Verify:  %.2f µs/op (%.0f ops/sec)%n",
                    totalVerify / 1000.0 / iterations,
                    iterations * 1_000_000_000.0 / totalVerify);

            // Clean up
            keys.destroy();
            device.close();

        } catch (QuacException e) {
            System.err.println("QUAC Error [" + e.getErrorCode() + "]: " + e.getMessage());
            e.printStackTrace();
        }

        System.out.println("\n=== Example Complete ===");
    }
}